Python/pylifecycle.c

```diff
@@ -1348,6 +1351,38 @@ Py_FinalizeEx(void)
     PyThreadState *tstate = _PyRuntimeState_GetThreadState(runtime);
     PyInterpreterState *interp = tstate->interp;
 
+    /* Check we're running in the main interpreter (not yet supported to call
+     * from any interpreter).
+     */
+    if (interp != PyInterpreterState_Main()) {
+        fprintf(stderr,
+                "Py_FinalizeEx: error: must be called from the main interpreter\n");
+        return -1;
+    }
+
+    // Finalize sub-interpreters.
+    runtime->interpreters.allow_new = 0;
+    PyInterpreterState *curr_interp = PyInterpreterState_Head();
+    PyInterpreterState *next_interp;
+    int64_t num_destroyed = 0;
+    while (curr_interp != NULL) {
+        next_interp = PyInterpreterState_Next(curr_interp);
+        if (curr_interp != interp) {
+            PyThreadState_Swap(curr_interp->tstate_head);
+            Py_EndInterpreter(curr_interp->tstate_head);
+            num_destroyed++;
+        }
+        curr_interp = next_interp;
+    }
+    PyThreadState_Swap(tstate);
+
+    if (num_destroyed > 0) {
+        if (PyErr_ResourceWarning(Py_None, 1,
+                                  "extra %zd interpreters", num_destroyed)) {
```

Comment:
You should mention that Python is being finalized and that these subinterpreters were still running.

For example, `subprocess.Popen.__del__()` says:

```
        if self.returncode is None:
            # Not reading subprocess exit status creates a zombie process which
            # is only destroyed at the parent python process exit
            _warn("subprocess %s is still running" % self.pid,
                  ResourceWarning, source=self)
```